### PR TITLE
Edx Oauth2 provider code flow and tests.

### PR DESCRIPTION
**Pull request by dominicfollett**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -156,14 +156,37 @@ RailsApiAuth.tap do |raa|
...
+user api, the username is need to retrieve user data, not just a valid
+oauth2 token. 
+
+E.g.
+
+```ruby
+headers = {
+	username: "alice", 
+	auth_code: "alices_authorization_code", 
+	grant_type: "edx_auth_code"
```
Comment (marcoow) [inline README.md]: never use tabs please

File: README.md [hunk 2]
```diff
@@ -156,14 +156,37 @@ RailsApiAuth.tap do |raa|
...
 end
 
 ```
 
+### A note on Edx Oauth2 code flows
+
+It is nesescary to include the Edx username in the when making a call
+rails_api_auth call /token. When rails_api_auth interfaces with Edx's
+user api, the username is need to retrieve user data, not just a valid
+oauth2 token.
```
Comment (marcoow) [inline README.md]: This sounds strange and if I'm not mistaken doesn't conform to the spec. Can you explain why that is?

Comment (dominicfollett) [inline README.md]: It is odd, however the Edx [user api](http://edx.readthedocs.io/projects/edx-platform-api/en/latest/user/overview.html#user-api-resources-and-endpoints) specified that a username is needed to retrieve user profile data. I'm not sure why they handled it this way. It certainly makes it less wieldy than the Google or Facebook method.

File: README.md (hunk 1 above)
Comment (dominicfollett) [inline README.md]: Noted, will update now.

**Comment by dominicfollett** [comment]
@marcoow I notice the CI builds are failing due to a dependency error:

```bash
An error occurred while installing ruby_dep (1.2.0), and Bundler cannot continue.
Make sure that `gem install ruby_dep -v '1.2.0'` succeeds before bundling.
```

I'm not sure how this might relate to my PR?

**Comment by marcoow** [comment]
@dominicfollett: rebase on `master` - I fixed the tests in #43

**Comment by dominicfollett** [comment]
@marcoow I'm getting test fails, due to a username being needed with Edx user API calls. I can fix this by providing my own tests to exercise this, I just wanted to check if you were ok with this, or perhaps want me to find another to get around this? I don't want to bloat your testing suite without due cause.

**Comment by marcoow** [comment]
@dominicfollett: can you point me to the change you're referring to?

**Comment by dominicfollett** [comment]
@marcoow I've added shared_examples/oauth2_edx_shared_requests.rb, where username needs to be set so it is passed as a parameter with the /token call. I also added a test to exercise the case where a username is not supplied.

**Comment by marcoow** [comment]
Can you revert all those file mode changes (644 -> 755)? I don't think those are correct.

**Comment by dominicfollett** [comment]
@marcoow I've updated that mode change. Apologies!

File: .travis.yml [hunk 3]
```diff
@@ -1,10 +1,9 @@
 language: ruby
 
 rvm:
-  - 2.0
-  - 2.1
-  - 2.2
-  - rbx-2
+  - 2.2.4
+  - 2.3.0
+  - jruby-9.0.5.0
```
Comment (marcoow) [inline .travis.yml]: this seems unrelated

File: app/controllers/oauth2_controller.rb [hunk 4]
```diff
@@ -16,6 +16,15 @@ def create
     when 'google_auth_code'
       authenticate_with_google(params[:auth_code])
     else
+      other_authenticators
```
Comment (marcoow) [inline app/controllers/oauth2_controller.rb]: Why do we handle edx and unknown authenticators in a separate method?

File: gemfiles/rails_3.gemfile [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 source "https://rubygems.org"
 
-gem "rails", "3.2.22"
+gem "rails", "3.2.22.2"
```
Comment (marcoow) [inline gemfiles/rails_3.gemfile]: this is unrelated actually

File: gemfiles/rails_4.gemfile [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 source "https://rubygems.org"
 
-gem "rails", "4.2.3"
+gem "rails", "4.2.6"
```
Comment (marcoow) [inline gemfiles/rails_4.gemfile]: this is unrelated actually

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -2,8 +2,10 @@
 
 require 'simplecov'
 SimpleCov.start do ||
-  minimum_coverage 95
-  refuse_coverage_drop
+  unless defined?(JRUBY_VERSION) || defined?(JRuby)
+    minimum_coverage 95
+    refuse_coverage_drop
+  end
```
Comment (marcoow) [inline spec/spec_helper.rb]: this is unrelated

**Comment by marcoow** [comment]
Cool, there's still a few file mode changes left and also some unrelated changes.

File: app/controllers/oauth2_controller.rb (hunk 4 above)
Comment (dominicfollett) [inline app/controllers/oauth2_controller.rb]: The Travis CI complains about method lengths longer than 10 lines.

**Comment by dominicfollett** [comment]
Thanks I'll update and clean up those I missed.

File: app/controllers/oauth2_controller.rb (hunk 4 above)
Comment (marcoow) [inline app/controllers/oauth2_controller.rb]: Can you disable that check for this specific method. Splitting it up like this might make Rubocop happy but actually makes the code much worse ;)

Comment (dominicfollett) [inline app/controllers/oauth2_controller.rb]: Ok sure thing :)

**Comment by dominicfollett** [comment]
@marcoow I think we're good to go? :)

File: .rubocop.yml [hunk 8]
```diff
@@ -44,6 +44,9 @@ Style/Documentation:
 Style/GuardClause:
   Enabled: false
 
+Metrics/MethodLength:
+  Enabled: false
+
```
Comment (marcoow) [inline .rubocop.yml]: Can you only disable this guard for the one method where it's failing?

Comment (dominicfollett) [inline .rubocop.yml]: I see, yeah that would be much better! Updated.

File: app/controllers/oauth2_controller.rb [hunk 9]
```diff
@@ -7,6 +7,7 @@ class Oauth2Controller < ApplicationController
 
   force_ssl if: -> { RailsApiAuth.force_ssl }
 
+  # rubocop:disable all
```
Comment (marcoow) [inline app/controllers/oauth2_controller.rb]: Only the method length check should be disable and only for this specific method (enable it again after the method definition).

Comment (dominicfollett) [inline app/controllers/oauth2_controller.rb]: Thanks for your patience, I misunderstood the semantics of rubocop here.

**Comment by dominicfollett** [comment]
@marcoow Thanks for your patience, and for pointing out those issues. I've re enabled rubocop method length checks after that create method.

**Comment by marcoow** [comment]
🎉

File: README.md [hunk 10]
```diff
@@ -156,16 +156,39 @@ RailsApiAuth.tap do |raa|
...
-  raa.google_client_id = '<your Google client id>'
+  raa.google_client_id     = '<your Google client id>'
   raa.google_client_secret = '<your Google client secret>'
-  raa.google_redirect_uri = '<your app redirect uri>'
+  raa.google_redirect_uri  = '<your app redirect uri>'
+
+  # Edx configurations
+  raa.edx_client_id     = '<your Edx client id>'
+  raa.edx_client_secret = '<your Edx client secret>'
+  raa.edx_domain        = '<your Edx app domain>'
```
Comment (marcoow) [inline README.md]: Why do we need to customize this? Does Edx give every app its own API domain?
